recursion/src/circuit_verifier.rs

```diff
@@ -0,0 +1,308 @@
+use alloc::vec;
+use alloc::vec::Vec;
+
+use itertools::{Itertools, zip_eq};
+use p3_circuit::utils::ColumnsTargets;
+use p3_circuit::{CircuitBuilder, ExprId};
+use p3_commit::Pcs;
+use p3_field::{BasedVectorSpace, Field, PrimeCharacteristicRing};
+use p3_uni_stark::StarkGenericConfig;
+
+use crate::recursive_traits::{
+    CommitmentTargets, OpenedValuesTargets, ProofTargets, Recursive, RecursiveAir, RecursivePcs,
+};
+
+#[derive(Debug)]
+pub enum VerificationError {
+    InvalidProofShape,
+    RandomizationError,
+}
+
+impl core::fmt::Display for VerificationError {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        match self {
+            VerificationError::InvalidProofShape => write!(f, "Invalid proof shape"),
+            VerificationError::RandomizationError => write!(
+                f,
+                "Missing random opened values for existing random commitment"
+            ),
+        }
+    }
+}
+
+// Method to get all the challenge wires.
+fn get_circuit_challenges<
+    SC: StarkGenericConfig,
+    Comm: Recursive<SC::Challenge, Input = <SC::Pcs as Pcs<SC::Challenge, SC::Challenger>>::Commitment>,
+    InputProof: Recursive<SC::Challenge>,
+    OpeningProof: Recursive<SC::Challenge>,
+    const D: usize,
+>(
+    proof_targets: &ProofTargets<SC, Comm, OpeningProof>,
+    circuit: &mut CircuitBuilder<SC::Challenge>,
+) -> Vec<ExprId>
+where
+    SC::Pcs: RecursivePcs<
+            SC,
+            InputProof,
+            OpeningProof,
+            Comm,
+            <SC::Pcs as Pcs<SC::Challenge, SC::Challenger>>::Domain,
+        >,
+{
+    let mut challenges = vec![];
+    // Observe degree bits and degree_bits - is_zk.
+    // Observe local wires.
+    // Observe public values.
+    challenges.push(circuit.add_public_input());
+    // Observe quotient chunks.
+    // Observe random commitment if any.
```

Comment:
can these comments be changed to TODOs so we can easily retrieve them and not forget them?